### PR TITLE
mac-virtualcam: Add macOS camera extension for macOS 13+

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
Adds native macOS camera extension as a virtual camera for macOS 13+.

### Motivation and Context
CMIO Camera Extensions allow developers to add virtual cameras (or USB camera drivers) to macOS in a secure way:

* Camera extensions need to be shipped with a codesigned and notarised app bundle
* Camera extensions are checked and loaded by the operating system, thus the virtual camera is available in all applications
* No 3rd party code needs to be loaded by host applications
* Camera extensions are automatically updated when the app is updated (and only updated if necessary)
* Users have to consent before a camera extension is installed
* Camera extensions run sandboxed

This also requires certain changes to the way OBS is built and run:

* Building OBS with a camera extensions requires(!) codesigning and a valid provisioning profile, as well as the entitlement for system extensions
* OBS needs to be run from the `/Applications` directory, macOS will not(!) load extensions from applications in different locations
* When not compiled on your own machine, the app also needs to be notarised

To avoid unnecessary hurdles when working on the project, the camera extension is *disabled* if the triplet of necessary UUIDs is not provided - no provisioning profile is needed to build and run OBS then, but the functionality is disabled.

**The current DAL plugin functionality used for virtual cameras was deprecated in macOS 12.3 and will be removed in macOS 14!**

Thus we need to have the "new" camera extension ready and released before that. Obvious benefit is that the virtual camera will appear in Facetime/QuickTime/Zoom/Discord/et. al. without workarounds or security issues.

### How Has This Been Tested?
Camera Extension has been tested on macOS 13 locally, build functionality on CI needs to be tested on CI, hence the draft status.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by PatTheMav** [comment]
While the code itself works, getting this to build correctly on CI has become a nightmare, as a working build needs to be code signed, notarised, have the right provisioning profile, and all associated identifiers being correct.

Allowing for "easy" local builds (without requiring any developer to have a paid developer account) and quick Ninja-based builds on CI will require re-jigging the entire macOS build system - using Xcode will do most things automatically for us, but we are not able to compose a full app bundle from within Xcode yet (without any fixups), so we need to remove all those manual fixups first).

I'll leave this in draft until I created/merged the required build system updates.

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:38]: ^ Available and deprecated since macOS 13, for our purposes

File: plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: What do you think about renaming `VIRTUALCAM_STREAM_UUID` to `VIRTUALCAM_SOURCE_UUID`?
I made such a change on my branch, probably after you looked at it. Both source and sink are streams so source_uuid would be more precise I think.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:39]: Personally not a great fan of this, can't fully put into words why. The Info.plist doesn't quite feel right for this kind of attribute. I had solved it by generating a C-Style header and using it as a bridgingheader, but I'm not sure that's optimal either.

Can you think of *some* other way? The answer might be "no" in the end, but maybe you can think of something else that I don't.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:26]: Crediting Jim with the copyright on this code is certainly nice to him but not correct I think, given that it's the you and me who wrote it. Maybe something more generic like "OBS Project" would be more correct.

File: plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: Co-authors need to be exactly `Co-authored-by: NAME <NAME@EXAMPLE.COM>` for GitHub to recognize them (also for the other commit).
For this commit specifically I wouldn't mind being the primary author either :P (though I can't quite tell how significantly you changed it and it probably doesn't matter too much).

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:17]: With the deployment target set to 13.0, are the availability checks needed?

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:115]: Nit: `LOG_ERROR` is meant for "places that should definitely not fail". Failure in these places is kinda expected, so I'd go for `LOG_WARNING` instead.

**Review by gxalpha — COMMENTED** [review]
Works as advertised once you learn the quirks with cmake 3.0. Left a few minor notes on the commit.

One thing I would also add is to default `ENABLE_VIRTUALCAM` to `OFF`, as it requires a provisioning now that many people who just want to build OBS don't have.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:39]: The `Info.plist` is to my knowledge actually the canonical way to put such information - this is where you add which services your application provides, which file types you support, which extensions you hold, etc. pp.

In essence, any meta "information" about your application should go into that file and the Cocoa functions exist for that reason - it's an easy way to read such information (which is also why we store information for Sparkle in there).

The plist itself cannot be changed after bundling/codesigning without invalidating the signature.

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:115]: `OSSystemExtensionErrorRequestCanceled` is not an error per-se because it is based on user interaction.

`OSSystemExtensionErrorUnsupportedParentBundleLocation` is an error, because you cannot use system extensions outside of the Applications directory.

Anything else is an unhandled situation and as such should always be treated as an error, because "you should usually not arrive at this point" and if you do it warrants the severity.

File: plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: The UUID is used for a `CMIOExtensionStreamSource`, so it's important to keep the name of the UUIDs matched with their purpose. Calling it a "source" UUID removes that logical connection.

Comment (gxalpha) [reply plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: In that case, both should contain `_STREAM_`, no? Both are just streams for the same device on different indexes (0 for the source stream and 1 for the sink stream) and directions.

Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: Err, you are right and I'm wrong - if you align it with the Swift classes, it should be `SOURCE` then.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (gxalpha) [reply plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:39]: Ah, I see 👍

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [reply plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:115]: > `OSSystemExtensionErrorUnsupportedParentBundleLocation` is an error, because you cannot use system extensions outside of the Applications directory.

What I'm referring to is the definition of purely the definition of `LOG_ERROR`, not necessarily an error in the sense of "something with the extension went wrong".

> Use if there's a problem that can potentially affect the program,
> but isn't enough to require termination of the program.
> 
> Use in creation functions and core subsystem functions.  Places that
> should definitely not fail.

I would argue that the fact that we're gracefully handling this, going as far as giving a localized response to it, makes it an expected-to-happen state (especially since every user will run into it at least once, until they allowed the extension in system settings).

> Anything else is an unhandled situation and as such should always be treated as an error, because "you should usually not arrive at this point" and if you do it warrants the severity.

Yeah you're right about these, anything else (wrong code signature etc) definitely shouldn't happen.

File: plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/CMakeLists.txt:1]: I just made you the author of the first two commits.

**Comment by PatTheMav** [comment]
> Works as advertised once you learn the quirks with cmake 3.0. Left a few minor notes on the commit.
> 
> One thing I would also add is to default `ENABLE_VIRTUALCAM` to `OFF`, as it requires a provisioning now that many people who just want to build OBS don't have.

It doesn't require one if you build with automatic codesigning in Xcode, though I'm not sure if this requires an Apple developer membership to work.

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [reply plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:115]: You convinced me about the default branch being something that definitely shouldn't happen, I'd agree with your initial assessment about it being error-worthy.

**Comment by gxalpha** [comment]
>> One thing I would also add is to default ENABLE_VIRTUALCAM to OFF, as it requires a provisioning now that many people who just want to build OBS don't have.
>
>It doesn't require one if you build with automatic codesigning in Xcode, though I'm not sure if this requires an Apple developer membership to work.

For the record: Resolved off-thread, the module of course gets disabled if you don't pass the required UUID's, which effectively makes it the default. I had passed the UUID's so it was turned on, which makes sense.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:17]: Could you try running it without the guard in a macOS 12 VM?

Comment (gxalpha) [reply plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:17]: Negative, can't log into iCloud on a VM (that's apparently normal, but means I can't build the app with entitlements) and my VM is unable to boot into recovery for some reason ("Internal Virtualization Error"). Would have to install macOS 12 genuinely.

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:98]: Explicit casts are redundant here I think, the `LOG_severity` enum values should be ints already

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:366]: The space in between `%s` and `%@` causes the new line to start the a space

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (gxalpha) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:41]: These look like a leftover of the code template for the camera extension that this uses which I didn't notice but aren't needed anymore.

File: plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in:28]: Indention is inconsistent here

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (gxalpha) [inline plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:24]: I wonder if this also needs some kind of `#if CAMERA_EXTENSION_AVAILABLE` define in addition to the runtime check, because as I understand (and correct me if I'm wrong) one is still able to compile OBS themselves without the camera extension(?). In such a case it would try to install/use an extension that doesn't exists.

**Review by gxalpha — COMMENTED** [review]
A few nitpicks (one of which more of a question), and commit authorship seems to have reset in one of the rebases.
Looks good otherwise I think! (or at least the program/extension code does, I'm illiterate when it comes to CI and cmake so can't review any of that :p)

File: plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/obs-plugin/plugin-main.mm:24]: In that case one has to configure the project with `ENABLE_VIRTUALCAM` set to `FALSE` to disable that functionality entirely.

Anyone with a valid Apple Developer certificate (or being a member of an Apple Developer team) can compile obs-studio with the extension and also run the build locally (no SIP disabling necessary). The only limitation is generating a signed build for distribution (the "Manual" code singing style), which requires changing the bundle identifier to a different value for which one has a valid provisioning profile.

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:172]: Any particular reason to use sha1 here over other options? Just no need for other options?

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:204]: Kind of feel like `provisionProfileUUID` would be more readable, but feel free to ignore if the refactor/rename would be annoying.

File: .github/workflows/main.yml
Comment (RytoEX) [inline .github/workflows/main.yml:536]: This appears to be a different method of creating a random password than above. Why two different methods?

File: UI/cmake/macos/entitlements-extension.plist
Comment (RytoEX) [inline UI/cmake/macos/entitlements-extension.plist:18]: Are these keys in any particular order?

File: cmake/macos/xcode.cmake
Comment (RytoEX) [inline cmake/macos/xcode.cmake:70]: Is this somehow different from 5.0? Wondering if the comment is still accurate.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:6]: Personally, I prefer ISO dates:
* 2022-09-30
* 2022-10-16

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:15]: Is the frame rate locked to 60 FPS?

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraProviderSource.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraProviderSource.swift:6]: Nit: Date format.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraStreamSink.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraStreamSink.swift:6]: Nit: date format.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:3]: Is "mac-camera-extension" appropriate here, or should this be "camera-extension"? Happy to defer to you.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraStreamSource.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/OBSCameraStreamSource.swift:6]: Nit: date format

File: plugins/mac-virtualcam/src/camera-extension/main.swift
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/main.swift:6]: Nit: date format

File: plugins/mac-virtualcam/src/obs-plugin/CMakeLists.txt
Comment (RytoEX) [inline plugins/mac-virtualcam/src/obs-plugin/CMakeLists.txt:26]: This seems oddly formatted.

File: plugins/mac-virtualcam/src/obs-plugin/data/locale/en-US.ini
Comment (RytoEX) [inline plugins/mac-virtualcam/src/obs-plugin/data/locale/en-US.ini:2]:
```suggestion
Error.SystemExtension.NotInstalled="The virtual camera is not installed.\n\nPlease allow OBS to install system software in System Settings → Privacy & Security → Security.\n\nYou may need to restart OBS if this message still appears afterward."
```

**Review by gxalpha — APPROVED** [review]

File: plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in
Comment (RytoEX) [inline plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in:26]: Is this change correct?

**Review by RytoEX — CHANGES_REQUESTED** [review]
A quick visual pass of the code. Will try to get a local build working soon.

File: plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in
Comment (gxalpha) [reply plugins/mac-virtualcam/src/obs-plugin/cmake/macos/Info.plist.in:26]: For context, on an earlier version of this PR on the Info.plist in the `camera-extension` directory I asked for this change (and I assume it then got changed for both files).
On this Info.plist here (in `obs-plugin`) I think neither the old nor the new is technically correct, it should probably be along the lines of `2020-${CURRENT_YEAR} John Boiles, Patrick Heyer, Sebastian Beckmann`.

File: .github/workflows/main.yml
Comment (PatTheMav) [reply .github/workflows/main.yml:172]: Just speed - it's mostly irrelevant as the password is per-session only.

File: .github/workflows/main.yml
Comment (PatTheMav) [reply .github/workflows/main.yml:536]: Mainly code evolution - at some point reading from `urandom` broke - should use the same variant as the code above.

File: UI/cmake/macos/entitlements-extension.plist
Comment (PatTheMav) [reply UI/cmake/macos/entitlements-extension.plist:18]: No order, as they're usually added automatically by Xcode.

File: cmake/macos/xcode.cmake
Comment (PatTheMav) [reply cmake/macos/xcode.cmake:70]: That's a mistake - `5.0` is the value accepted by Xcode, when specifying `5` it's treated as a custom (and ignored) setting.

File: plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/OBSCameraDeviceSource.swift:15]: Correct, the device (as exposed to other applications) has a single resolution/framerate configuration - the incoming pixel buffers are converted automatically.

File: plugins/mac-virtualcam/src/obs-plugin/CMakeLists.txt
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/obs-plugin/CMakeLists.txt:26]: That's because `cmake-format` can't handle custom functions with sets of tuples as variables, needs to be manually formatted.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (RytoEX) [inline plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:14]: These strings still say `mac-camera-extension`.

**Review by RytoEX — CHANGES_REQUESTED** [review]
One question about a string.

I tested locally on macOS 13.3.1 on my M2 MacBook Pro and it worked fine in Discord, Chrome, Firefox, FaceTime, Webcamoid, and Zoom.

File: plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in
Comment (PatTheMav) [reply plugins/mac-virtualcam/src/camera-extension/cmake/macos/Info.plist.in:14]: That is indeed on purpose.

**Comment by MisutaaAsriel** [comment]
Sorry for asking this way, but is there a way to test this plugin *before* it becomes part of an official build? When I try builds created by GitHub actions it just complains about signing/entitlements.

**Comment by PatTheMav** [comment]
> @MisutaaAsriel I am surprised that GitHub builds even contain the camera extension, they weren't initially, I believe this changed in one of the later CI script rewrites. You're correct though, builds from GitHub Actions do not work as-is (because they are not codesigned or notarized).

Nightly builds run on master branch, so the provisioning profile is available to them and the feature is mainly guarded behind the availability of that. We could notarise them if we feel confident that our master branch always contains "safe" code.

**Comment by felippenardi** [comment]
The new com.obsproject.obs-studio.mac-camera-extension is using a LOT of CPU usage. Wasn't that observed before?

**Pull request merged**
